### PR TITLE
gssapi: default client keytab (#383)

### PR DESCRIPTION
**Pull request by lhoward**
WIP for #383 – not sure if fix should be this simple?

### CODE REVIEW

File: lib/gssapi/krb5/acquire_cred.c [hunk 1]
```diff
@@ -80,6 +80,34 @@ get_keytab(krb5_context context, krb5_keytab *keytab)
...
+static krb5_error_code
+get_client_keytab(krb5_context context,
+		  krb5_const_principal principal,
+		  krb5_keytab *keytab)
+{
+    krb5_error_code kret;
+    krb5_keytab_entry entry;
+    char *name = NULL;
+
+    kret = _krb5_kt_client_default_name(context, &name);
```
Comment (nicowilliams) [inline lib/gssapi/krb5/acquire_cred.c]: Not a required change, but you can deindent a lot like so:

```
    kret = _krb5_kt_client_default_name(context, &name);
    if (kret == 0)
        kret = krb5_kt_resolve(context, name, keytab);
    if (kret == 0 && principal != NULL)
	kret = krb5_kt_get_entry(context, *keytab, principal,
				 0, 0, &entry);
    if (kret == 0)
	krb5_kt_free_entry(context, &entry);
    krb5_xfree(name);
    if (kret == 0)
        return 0;

    return get_keytab(context, keytab); /* fall back on regular keytab */
```

File: lib/gssapi/krb5/acquire_cred.c [hunk 2]
```diff
@@ -80,6 +80,34 @@ get_keytab(krb5_context context, krb5_keytab *keytab)
...
+    if (kret == 0 && principal != NULL) {
+	kret = krb5_kt_get_entry(context, *keytab, principal,
+				 0, 0, &entry);
+	if (kret == 0)
+	    krb5_kt_free_entry(context, &entry);
+    }
+    krb5_xfree(name);
+
+    if (kret != 0)
+	kret = get_keytab(context, keytab);
```
Comment (nicowilliams) [inline lib/gssapi/krb5/acquire_cred.c]: If a principal name was given, do we not want to check if this keytab has entries for it?

File: lib/gssapi/krb5/acquire_cred.c (hunk 1 above)
Comment (lhoward) [reply lib/gssapi/krb5/acquire_cred.c]: Yeah, thought about that, wasn't convinced it was necessarily more readable. But happy to.

File: lib/gssapi/krb5/acquire_cred.c (hunk 2 above)
Comment (lhoward) [reply lib/gssapi/krb5/acquire_cred.c]: Probably but – didn't want to change current behavior. Hmm, does Heimdal do the MIT thing where rd_req tries all the keys in the keytab? We'd have to be careful we don't break that (if it does that).

Comment (lhoward) [reply lib/gssapi/krb5/acquire_cred.c]: Don't think it'll make much difference though for the client, it'd just fail later on.

Comment (lhoward) [reply lib/gssapi/krb5/acquire_cred.c]: Hmm I guess we don’t handle the case where we don’t know the client principal name. Todo.

**Comment by lhoward** [comment]
@nicowilliams the only thing with this patch (well, with the existing Heimdal code), is that if you don't specify a principal when acquiring a GSS credential, it uses `krb5_get_default_principal()`. Arguably if there's a client keytab, we should use the first principal we find there. Or perhaps not. Anyway, if you think this is a worthwhile change I can make it.

**Comment by lhoward** [comment]
Excerpt below, see with `get_client_keytab()` I've just (lazily!) hooked into the existing keytab path.

```c
try_keytab:
    if (handle->principal == NULL) {
        /* We need to know what client principal to use */
        kret = krb5_get_default_principal(context, &handle->principal);
        if (kret)
            goto end;
    }
    kret = get_client_keytab(context, handle->principal, &keytab);
    if (kret)
        goto end;
```

**Comment by lhoward** [comment]
But there's no canonical principal for a keytab so...

**Comment by nicowilliams** [comment]
@lhoward It's very tricky...

kinit with a keytab (or `gss_acquire_cred()` with a keytab) and no other source of a principal name can do things like:

- try `${USER}` as the principal name
- try the first/last keytab entry for a principal that has one component
- try the first/last keytab entry

I say "last" mostly because keytabs get appended to, and rarely get re-sorted.  This is a very obnoxious thing about keytabs...

Ideally things like kadmin's `ext_keytab` and ktutil should re-sort the keytab to put the newest keytab entries first, which helps performance of `krb5_rd_req*()` and `gss_accept_sec_context()`!  But since the order in which we try things is meaningful (see above), this strikes me as a notable behavior change, perhaps even a backwards-incompatible behavior change.

One way to pick default keytab entries that is less dependent on keytab entry order is to pick those with the highest set time (fortunately keytab entries record the time at which they were created).  Here keytab entry order matters only when you have multiple entries with the same newest set-time, and that's unlikely to happen for the same principal name and kvno.

What a pain.

**Pull request merged**